unison-cli/src/Unison/Cli/MonadUtils.hs

```diff
@@ -329,167 +342,138 @@ assertNoBranchAtPath' path' = do
+stepManyAtM pb reason actions = do
+  updateProjectBranchRoot pb reason \oldRoot -> do
+    newRoot <- liftIO (Branch.stepManyAtM (makeActionsUnabsolute actions) oldRoot)
+    pure (newRoot, ())
 
 -- | Update a branch at the given path, returning `True` if
 -- an update occurred and false otherwise
 updateAtM ::
   Text ->
-  Path.Absolute ->
+  ProjectPath ->
   (Branch IO -> Cli (Branch IO)) ->
   Cli Bool
-updateAtM reason (Path.Absolute p) f = do
-  b <- getRootBranch
-  b' <- Branch.modifyAtM p f b
-  updateRoot b' reason
-  pure $ b /= b'
+updateAtM reason pp f = do
+  oldRootBranch <- getProjectBranchRoot (pp ^. #branch)
+  newRootBranch <- Branch.modifyAtM (pp ^. PP.path_) f oldRootBranch
+  updateProjectBranchRoot_ (pp ^. #branch) reason (const newRootBranch)
+  pure $ oldRootBranch /= newRootBranch
 
 -- | Update a branch at the given path, returning `True` if
 -- an update occurred and false otherwise
 updateAt ::
   Text ->
-  Path.Absolute ->
+  ProjectPath ->
   (Branch IO -> Branch IO) ->
   Cli Bool
-updateAt reason p f = do
-  updateAtM reason p (pure . f)
+updateAt reason pp f = do
+  updateAtM reason pp (pure . f)
 
 updateAndStepAt ::
-  (Foldable f, Foldable g) =>
+  (Foldable f, Foldable g, Functor g) =>
   Text ->
+  ProjectBranch ->
   f (Path.Absolute, Branch IO -> Branch IO) ->
-  g (Path, Branch0 IO -> Branch0 IO) ->
+  g (Path.Absolute, Branch0 IO -> Branch0 IO) ->
   Cli ()
-updateAndStepAt reason updates steps = do
-  root <-
-    (Branch.stepManyAt steps)
-      . (\root -> foldl' (\b (Path.Absolute p, f) -> Branch.modifyAt p f b) root updates)
-      <$> getRootBranch
-  updateRoot root reason
-
-updateRoot :: Branch IO -> Text -> Cli ()
-updateRoot new reason =
-  Cli.time "updateRoot" do
-    Cli.Env {codebase} <- ask
-    let newHash = Branch.headHash new
-    oldHash <- getLastSavedRootHash
-    when (oldHash /= newHash) do
-      liftIO (Codebase.putRootBranch codebase reason new)
-      setRootBranch new
-      setLastSavedRootHash newHash
+updateAndStepAt reason projectBranch updates steps = do
+  let f b =
+        b
+          & (\root -> foldl' (\b (Path.Absolute p, f) -> Branch.modifyAt p f b) root updates)
+          & (Branch.stepManyAt (first Path.unabsolute <$> steps))
+  updateProjectBranchRoot_ projectBranch reason f
+
+updateProjectBranchRoot :: ProjectBranch -> Text -> (Branch IO -> Cli (Branch IO, r)) -> Cli r
+updateProjectBranchRoot projectBranch reason f = do
+  Cli.Env {codebase} <- ask
+  Cli.time "updateProjectBranchRoot" do
+    old <- getProjectBranchRoot projectBranch
+    (new, result) <- f old
+    when (old /= new) do
+      liftIO $ Codebase.putBranch codebase new
+      Cli.runTransaction $ do
+        -- TODO: If we transactionally check that the project branch hasn't changed while we were computing the new
+        -- branch, and if it has, abort the transaction and return an error, then we can
+        -- remove the single UCM per codebase restriction.
```

Comment:
It would be nice to do this at some point, but we can do it later, this PR already improves our atomicity by a lot.